src/annotator/anchoring/pdf.js

```diff
@@ -446,57 +446,71 @@ function prioritizePages(position) {
 /**
  * Anchor a set of selectors to a DOM Range.
  *
+ * `selectors` must include a `TextQuoteSelector` and may include other selector
+ * types.
+ *
  * @param {HTMLElement} root
- * @param {Selector[]} selectors - Selector objects to anchor
+ * @param {Selector[]} selectors
  * @return {Promise<Range>}
  */
-export function anchor(root, selectors) {
-  const position = /** @type {TextPositionSelector|undefined} */ (
-    selectors.find(s => s.type === 'TextPositionSelector')
-  );
+export async function anchor(root, selectors) {
   const quote = /** @type {TextQuoteSelector|undefined} */ (
     selectors.find(s => s.type === 'TextQuoteSelector')
   );
 
-  /** @type {Promise<Range>} */
-  let result = Promise.reject('unable to anchor');
+  // The quote selector is required in order to check that text position
+  // selector results are still valid.
+  if (!quote) {
+    throw new Error('No quote selector found');
+  }
+
+  const position = /** @type {TextPositionSelector|undefined} */ (
+    selectors.find(s => s.type === 'TextPositionSelector')
+  );
 
   if (position) {
-    result = result.catch(() => {
-      return findPage(position.start).then(({ index, offset, textContent }) => {
-        const start = position.start - offset;
-        const end = position.end - offset;
-        const length = end - start;
-
-        const matchedText = textContent.substr(start, length);
-        if (quote && quote.exact !== matchedText) {
-          throw new Error('quote mismatch');
-        }
+    // If we have a position selector, try using that first as it is the fastest
+    // anchoring method.
+    try {
+      const { index, offset, textContent } = await findPage(position.start);
+      const start = position.start - offset;
+      const end = position.end - offset;
+      const length = end - start;
+
+      const matchedText = textContent.substr(start, length);
+      if (quote.exact !== matchedText) {
+        throw new Error('quote mismatch');
+      }
 
-        return anchorByPosition(index, start, end);
-      });
-    });
-  }
+      const range = await anchorByPosition(index, start, end);
+      return range;
+    } catch {
+      // Fall back to quote selector
+    }
 
-  if (quote) {
-    result = result.catch(() => {
+    // If anchoring with the position failed, check for a cached quote-based
+    // match using the quote + position as a cache key.
+    try {
       if (
-        position &&
         quotePositionCache[quote.exact] &&
         quotePositionCache[quote.exact][position.start]
       ) {
         const { pageIndex, anchor } =
           quotePositionCache[quote.exact][position.start];
-        return anchorByPosition(pageIndex, anchor.start, anchor.end);
+        const range = await anchorByPosition(
```

Comment:
```suggestion
        return anchorByPosition(
```

This matches the style of the last return statement in this function.

Reply:
There's a subtle problem with that! Dropping the `await` means that any errors thrown by `anchorByPosition` will no longer cause control to move to the `catch` block below. Changing the code to `return await anchorByPosition(...)` would work, but I thought naming the result was helpful to the reader.